README.md
# nginx-auth-server

A lightweight authentication server designed to be used in conjunction with nginx 'http_auth_request_module'. nginx-auth-server provides an additional authentication layer that is useful for reverse proxy scenarios, where the proxied service does not support user authentication.

## Table of Contents

- [Demo](#demo)
- [Features](#features)
- [Getting Started](#getting-started)
  - [With Docker](#with-docker)
  - [Native](#native)
- [Contributing](#contributing)
- [Documentation](#documentation)
- [Changelog](#changelog)
- [Credits](#credits)
- [License](#license)

## Demo

![demo.gif](https://raw.githubusercontent.com/burakkavak/nginx-auth-server/master/assets/demo.gif)

## Features

- low latency (<1ms)
- support for Two-Factor Authentication (2FA)
- support for LDAP to validate user credentials
- optional bot protection with Google reCAPTCHA
- location logging with MaxMind® GeoIP2 database

## Getting Started

### With Docker

Download the [docker-compose.yml](https://raw.githubusercontent.com/burakkavak/nginx-auth-server/master/docker-compose.yml) into a directory of your liking.
```shell
$ mkdir -p ~/docker/nginx-auth-server
$ cd ~/docker/nginx-auth-server
$ wget --content-disposition https://raw.githubusercontent.com/burakkavak/nginx-auth-server/master/docker-compose.yml
```

Start the container:
```shell
$ docker-compose up -d
```

You can now point a NGINX server to this docker container, please refer to the '[Native](#native)' section for a NGINX configuration example.

The CLI is called using *docker exec*. Please refer to the [CLI reference](https://burakkavak.github.io/nginx-auth-server/#command-line-interface-cli) for all commands. Here are some examples:
```shell
# docker exec -it <container_name> nginx-auth-server <command_parameters>
$ docker exec -it nginx-auth-server nginx-auth-server user add --username foo
$ docker exec -it nginx-auth-server nginx-auth-server user list
$ docker exec -it nginx-auth-server nginx-auth-server cookie list
```

#### Environment variables

The docker application can be configured using environment variables. Modify the *docker-compose.yml* and restart the container so the changes take effect.

| Environment variable | Default value | Description |
|:-:|:-:|--|
| `SERVER_LISTEN_ADDRESS` | `0.0.0.0` | The HTTP(S) server is listening to requests on this address (inside the container) |
| `SERVER_LISTEN_PORT` | `17397` | The application is going to listen for HTTP requests on this port (inside the container) |
| `SERVER_DOMAIN` | `localhost` | Domain used to set the authentication cookie and as issuer for TOTP. E.g. `example.org` |
| `TLS_ENABLED` | `false` | Enable HTTPS/TLS encryption for the webserver. The unencrypted HTTP server will be disabled |
| `TLS_LISTEN_PORT` | `17760` | The application is going to listen for HTTPS requests on this port (inside the container) |
| `TLS_CERT_PATH` | `/opt/nginx-auth-server/certs/server.crt` | Path of the SSL certificate (inside the container) |
| `TLS_CERT_KEY` | `/opt/nginx-auth-server/certs/server.key` | Path of the SSL certificate key (inside the container) |
| `COOKIES_LIFETIME` | `7` | Cookie lifetime in days. User has to re-authenticate after expiration |
| `COOKIES_SECURE` | `true` | Set secure attribute for cookies. The browser will only send the auth cookie in a HTTPS context if this is enabled |
| `COOKIES_NAME` | `Nginx-Auth-Server-Token` | Set cookie name attribute for cookie. Useful if you are running more than one auth server on the same domain. The cookie name can contain any US-ASCII characters, but some exception apply (refer to https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie) |
| `LDAP_ENABLED` | `false` | Enable/disable LDAP support. The application will prioritize local authentication data first |
| `LDAP_URL` |  | LDAP url. Example for TLS connection: `ldaps://ldap.example.com:636`. Example for non-TLS connection: `ldap://ldap.example.com:389` |
| `LDAP_ORGANIZATIONAL_UNIT` | `users` | LDAP organizational unit (OU) that is used to search the user |
| `LDAP_DOMAIN_COMPONENTS` |  | LDAP baseDN (DC) of the LDAP tree. Example: `dc=example,dc=org` |
| `RECAPTCHA_ENABLED` | `false` | Enable/disable Google reCAPTCHA v2 (invisible) support for the login form |
| `RECAPTCHA_SITE_KEY` |  | reCAPTCHA site key that is provided by Google upon site creation |
| `RECAPTCHA_SECRET_KEY` |  | reCAPTCHA secret key that is provided by Google upon site creation |
| `GEOIP_PATH` | `` | File path to MaxMind® GeoLite2 City database (inside the container) |

### Native

Download the appropriate binary from the [Releases](https://github.com/burakkavak/nginx-auth-server/releases) section.

Download the current [config.ini](https://raw.githubusercontent.com/burakkavak/nginx-auth-server/master/config.ini) into the same directory:

```shell
$ wget --content-disposition https://raw.githubusercontent.com/burakkavak/nginx-auth-server/master/config.ini
```

Run the server:
```shell
$ ./nginx-auth-server run
```

For user management (adding/removing users) refer to the CLI usage information:
```shell
$ ./nginx-auth-server help
$ ./nginx-auth-server user add --username foo --otp
```

Reconfigure nginx server:
```nginx
server {
  listen 80 default_server;
  listen [::]:80 default_server;

  root /var/www/html;

  index index.html index.htm index.nginx-debian.html;

  server_name _;

  # Redirect user to /login if nginx-auth-server responds with '401 Unauthorized'
  error_page 401 /login;

  location / {
    auth_request /auth;

    # pass Set-Cookie headers from the subrequest response back to requestor
    auth_request_set $auth_cookie $upstream_http_set_cookie;
    add_header Set-Cookie $auth_cookie;

    auth_request_set $auth_status $upstream_status;

    # serve files if the user is authenticated
    try_files $uri $uri/ /index.html;
  }

  location = /auth {
    # internally only, /auth can not be accessed from outside
    internal;

    # nginx-auth-server running on port 17397
    proxy_pass http://localhost:17397;

    # don't pass request body to proxied server, we only need the headers which are passed on by default
    proxy_pass_request_body off;

    # there is no content length since we stripped the request body
    proxy_set_header Content-Length "";

    # let proxy server know more details of request
    proxy_set_header X-Original-URI $request_uri;
    proxy_set_header X-Original-Remote-Addr $remote_addr;
    proxy_set_header X-Original-Host $host;
  }

  # these are handled by nginx-auth-server as part of the auth routines
  location ~ ^/(login|logout|whoami)$ {
    proxy_pass http://localhost:17397;

    proxy_set_header X-Original-URI $request_uri;
    proxy_set_header X-Original-Remote-Addr $remote_addr;
    proxy_set_header X-Original-Host $host;
  }

  # static nginx-auth-server assets (css, js, ...)
  location /nginx-auth-server-static {
    proxy_pass http://localhost:17397/nginx-auth-server-static;

    proxy_set_header X-Original-URI $request_uri;
    proxy_set_header X-Original-Remote-Addr $remote_addr;
    proxy_set_header X-Original-Host $host;
  }
}
```

You can also run the server as a systemd service. Example configuration for user *www-data*:
```apacheconf
[Unit]
Description=nginx-auth-server
After=network.target

[Service]
Type=simple
User=www-data
Group=www-data
WorkingDirectory=/var/www/nginx-auth-server
ExecStart=/var/www/nginx-auth-server/nginx-auth-server run
Restart=on-failure
# Other restart options: always, on-abort, etc

# The install section is needed to use
# `systemctl enable` to start on boot
# For a user service that you want to enable
# and start automatically, use `default.target`
# For system level services, use `multi-user.target`
[Install]
WantedBy=multi-user.target
```

## Contributing

Fork this repo and checkout the *develop* branch.

```shell
$ git clone <your_forked_repo> -b develop
$ cd nginx-auth-server
```

Install the npm dependencies.

```shell
$ npm i
```

Build the JavaScript/TypeScript/SCSS stack once.

```shell
$ npm run build
```

Run the Go application

```shell
$ go build -o nginx-auth-server ./src/ && ./nginx-auth-server run
```

You can now point a nginx webserver to this auth-server. Refer to the nginx configuration in the *[Getting Started](#getting-started)* section.

If you want to make changes in the TypeScript/SCSS, you can run npm in *watch* mode:

```shell
$ npm run watch-ts
$ npm run watch-scss
```

**You have to restart the Go application after every change for the changes to take effect.**

## Documentation

The CLI and HTTP API documentation is available here: [https://burakkavak.github.io/nginx-auth-server/](https://burakkavak.github.io/nginx-auth-server/)

## Changelog

See [`CHANGELOG`](./CHANGELOG.md)

## Credits

- [etcd-io/bbolt](https://github.com/etcd-io/bbolt)
- [gin-gonic/gin](https://github.com/gin-gonic/gin)
- [go-ini/ini](https://github.com/go-ini/ini)
- [go-ldap/ldap](https://github.com/go-ldap/ldap)
- [oschwald/geoip2-golang](https://github.com/oschwald/geoip2-golang)
- [pquerna/otp](https://github.com/pquerna/otp)
- [urfave/cli](https://github.com/urfave/cli)

## License

See [`LICENSE`](./LICENSE)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.0.10] - TBD
- added support for changing the cookie name (see: https://github.com/burakkavak/nginx-auth-server/issues/3)

## [0.0.9] - 2023-03-23
- fixed IP address logging upon authentication to log the real client IP
- fixed JS/CSS caching problems with browsers
- */login* route now redirects to a callback URL instead of returning a 200 status code when the user is already authenticated
- updated dependencies (including an update to TypeScript 5 and Golang 1.20.x). this also should fix some security vulnerabilities with *golang.org/x/net*

## [0.0.8] - 2023-01-15
- performance optimization to mitigate argon2 performance hit
- added 'show password' functionality in the login form
- improved logging: application logs to 'auth.log' and 'app.log' now
- fixed long Docker container shutdown time
- fixed bug where any TOTP input was rejected

## [0.0.7] - 2023-01-11
- Docker support
- support for HTTPS/TLS
- password parameter removed from user creation to avoid bash/zsh entries with plain text passwords
- expired cookies are automatically deleted from the database and cache now
- check for existing (LDAP) users before user creation

## [0.0.6] - 2023-01-05
- updated security (*argon2* is being used instead of *bcrypt* now)
- implemented cache for better performance
- cookie lifetime and secure attribute is now customizable in config.ini
- upon cookie expiration, the user will get a notice in the frontend now
- updated Go to version 1.19
- updated Go dependencies
- updated Node.js dependencies

## [0.0.5] - 2022-08-25
- fixed an issue with the libc-dependency in the binaries, that prevented the application from running on older libc versions

## [0.0.4] - 2022-07-22
- added /whoami API endpoint

## [0.0.3] - 2022-07-22
- added LDAP support with [go-ldap/ldap](https://github.com/go-ldap/ldap)
- added Google reCAPTCHA v2 support for login form

## [0.0.2] - 2022-07-18
- added TOTP support with [pquerna/otp](https://github.com/pquerna/otp)
- added [QRencode](https://github.com/fukuchi/libqrencode) support to display generated TOTP secrets in terminal
- changed/improved frontend form validation
- remove associated user cookies upon user deletion
- improved cookie security

## [0.0.1] - 2022-07-17
- initial MVP implementation
- implemented frontend login form
- implemented authentication cookie
- implemented data persistence with [etcd-io/bbolt](https://github.com/etcd-io/bbolt)
- implemented configuration parsing with [go-ini/ini](https://github.com/go-ini/ini)
- implemented webserver with [gin-gonic/gin](https://github.com/gin-gonic/gin)
- implemented CLI with [urfave/cli](https://github.com/urfave/cli)


src/main.go
package main

import (
	"bytes"
	"context"
	"embed"
	"encoding/json"
	"errors"
	"fmt"
	"html/template"
	"io"
	"math/rand"
	"net/http"
	"os"
	"os/exec"
	"os/signal"
	"strconv"
	"syscall"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/pquerna/otp/totp"
)

// templateFiles contains any files that are served prefixed with the relative URL /nginx-auth-server-static.
// These files are embedded in the final executable using go:embed.
//
//go:embed templates
var templateFiles embed.FS

// staticFiles contains any files that are served prefixed with the relative URL /nginx-auth-server-static.
// These files are embedded in the final executable using go:embed.
//
//go:embed css js
var staticFiles embed.FS

// GinMode describes the Gin web framework operating mode. This variable is overwritten in prod.go
// if the executable is build using the Go build tag 'prod'.
var GinMode = gin.DebugMode

const AppVersion = "0.0.9"

func main() {
	rand.Seed(time.Now().UTC().UnixNano())

	if err := app.Run(os.Args); err != nil {
		fmt.Printf(err.Error())
	}
}

// runGin sets up the Gin router and starts the webserver.
func runGin() {
	gin.SetMode(GinMode)

	router := gin.Default()
	templates, _ := template.ParseFS(templateFiles, "templates/*.html")
	router.SetHTMLTemplate(templates)
	router.StaticFS("/nginx-auth-server-static", http.FS(staticFiles))

	router.GET("/auth", authenticate)
	router.GET("/login", login)
	router.POST("/login", processLoginForm)
	router.GET("/logout", logout)
	router.GET("/whoami", whoami)

	serverAddress := GetListenAddress() + ":" + strconv.Itoa(GetListenPort())
	tlsEnabled := GetTlsEnabled()
	tlsCertPath := GetTlsCertPath()
	tlsKeyPath := GetTlsKeyPath()

	// overwrite serverAddress if TLS is configured and enabled
	if tlsEnabled {
		serverAddress = GetListenAddress() + ":" + strconv.Itoa(GetTlsListenPort())

		if err := CheckFileReadable(tlsCertPath); err != nil {
			appLog.Fatalf("fatal error: TLS certificate at '%s' does not exist or is not readable. Check the configuration and/or file permissions.", GetTlsCertPath())
		}

		if err := CheckFileReadable(tlsKeyPath); err != nil {
			appLog.Fatalf("fatal error: TLS key at '%s' does not exist or is not readable. Check the configuration and/or file permissions.", GetTlsKeyPath())
		}
	}

	server := &http.Server{
		Addr:    serverAddress,
		Handler: router,
	}

	// start the webserver in HTTP or HTTPS mode
	go func() {
		var err error = nil

		if tlsEnabled {
			appLog.Printf("listening and serving HTTPS request on %s\n", serverAddress)
			err = server.ListenAndServeTLS(tlsCertPath, tlsKeyPath)
		} else {
			appLog.Printf("listening and serving HTTP request on %s\n", serverAddress)
			err = server.ListenAndServe()
		}

		if err != nil && err != http.ErrServerClosed {
			appLog.Fatalf("fatal error trying to launch the webserver: %s\n", err)
		}
	}()

	// gracefully quit Gin server (https://gin-gonic.com/docs/examples/graceful-restart-or-stop/)
	quit := make(chan os.Signal)
	signal.Notify(quit, syscall.SIGINT, syscall.SIGTERM)
	<-quit
	appLog.Println("Shutting down webserver...")

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	if err := server.Shutdown(ctx); err != nil {
		appLog.Fatalf("fatal error: could not shutdown server gracefully. %s\n", err)
	}
}

// addUser receives the username and plaintext password and adds the new user to the database.
// If the password is empty, addUser will generate a password.
func addUser(username string, password string, otp bool) {
	if username == "" {
		appLog.Fatalf("invalid username")
	}

	if GetUserByUsernameCaseInsensitive(username) != nil {
		appLog.Fatalf("user with username %s already exists", username)
	}

	// generate password if empty password was given
	if password == "" {
		generatedPassword := GeneratePassword(8, 1, 1)

		fmt.Printf("no password given, generated password for user '%s': '%s'\n", username, generatedPassword)

		addUser(username, generatedPassword, otp)
	} else if err := CheckPasswordRequirements(password); err != nil {
		fmt.Printf("password does not meet minimum requirements: %s\n", err)
		return
	} else {
		// hash password using argon2 for database storage
		encodedPasswordHash := GenerateHash(password)

		var encryptedOtpSecret []byte

		if otp {
			otpKey, err := totp.Generate(totp.GenerateOpts{
				Issuer:      GetDomain(),
				AccountName: username,
			})

			if err != nil {
				appLog.Fatalf("could not create TOTP: %s", err)
			}

			fmt.Printf("TOTP secret key for user '%s': '%s'\n", username, otpKey.Secret())

			// output TOTP url as QR code to stdout using libqrencode
			output, err := exec.Command("sh", "-c", fmt.Sprintf("qrencode -t UTF8 '%s'", otpKey.URL())).Output()

			fmt.Printf("TOTP URL for user '%s': '%s'\n", username, otpKey.URL())

			if err != nil {
				fmt.Println("install 'qrencode' library to display a QR code.")
			} else {
				fmt.Println(string(output))
			}

			// encrypt TOTP secret using user password for database storage
			encryptedOtpSecret = Encrypt([]byte(otpKey.Secret()), password)
		}

		user := User{
			Username:  username,
			Password:  encodedPasswordHash,
			OtpSecret: encryptedOtpSecret,
		}

		err = CreateUser(&user)

		if err != nil {
			appLog.Fatalf("fatal error: could not save user to database: %s", err)
		} else {
			appLog.Printf("user with username '%s' successfully created\n", username)
		}
	}
}

// removeUser removes a user from the database.
// TODO: don't delete associated user cookies if there is an existing LDAP user with the same username
func removeUser(username string) {
	err := RemoveUser(username)

	if err != nil {
		appLog.Fatalf("fatal error: could not remove user from database: %s", err)
	} else {
		appLog.Printf("user with username '%s' has been removed\n", username)
	}

	err = DeleteCookiesByUsername(username)

	if err != nil {
		appLog.Fatalf("fatal error: could not remove user associated cookies from database for username '%s': %s\n", username, err)
	} else {
		appLog.Printf("user associated cookies for username '%s' have been removed\n", username)
	}
}

// authenticate handles the /auth route. If a valid cookie is found in the request header, the
// the response will be 200. If the cookie is invalid or expired, 401 is set as a response status.
func authenticate(c *gin.Context) {
	token, err := c.Cookie(GetCookieName())

	if err != nil {
		c.AbortWithStatus(401)
		return
	}

	if _, err = VerifyCookie(token); err != nil {
		c.AbortWithStatus(401)
		return
	} else {
		c.Status(200)
		return
	}

}

// login handles the /login route. If a valid cookie is found in the request header, the
// the response will be 302 redirect to the given 'callback' query param. If no callback is given, the user
// will be redirected to the root page. If the user is not authenticated,
// the login form template will be displayed.
func login(c *gin.Context) {
	token, err := c.Cookie(GetCookieName())

	if err == nil {
		if _, err = VerifyCookie(token); err == nil {
			// user already authorized
			// refer to: https://github.com/burakkavak/nginx-auth-server/issues/2
			c.Redirect(302, c.Query("callback"))
			return
		}
	}

	// attach all embedded CSS/JS files to the HTML template
	cssFiles := GetFilenamesFromFS(staticFiles, "css")
	jsFiles := GetFilenamesFromFS(staticFiles, "js")

	c.HTML(http.StatusOK, "login.html", gin.H{
		"cssFiles":         cssFiles,
		"jsFiles":          jsFiles,
		"recaptchaEnabled": GetRecaptchaEnabled(),
		"recaptchaSiteKey": GetRecaptchaSiteKey(),
	})
}

// logout handles the /logout route. If a valid cookie is found in the request header, the
// the response will be 200 and the cookie will be deleted.
func logout(c *gin.Context) {
	token, err := c.Cookie(GetCookieName())
	clientIp := GetClientIpFromContext(c)

	if err != nil {
		c.AbortWithStatus(401)
		return
	}

	if cookie, err := VerifyCookie(token); err == nil {
		DeleteCookieFromCache(cookie)
		err := DeleteCookie(cookie)

		if err != nil {
			c.AbortWithError(http.StatusInternalServerError, errors.New("could not delete user cookie from database"))
			authLog.Printf("error: user with username '%s' and client IP '%s' tried logging out,"+
				"but the associated authentication cookie could not be deleted from the database. %s\n", cookie.Username, clientIp, err)
			return
		} else {
			http.SetCookie(c.Writer, &http.Cookie{
				Name:     cookie.Name,
				Value:    fmt.Sprintf("$username=%s,$value=%s", cookie.Username, cookie.Value),
				Expires:  time.Now(),
				Domain:   cookie.Domain,
				HttpOnly: cookie.HttpOnly,
				Secure:   cookie.Secure,
			})

			c.String(http.StatusOK, "user successfully logged out")
			authLog.Printf("user with username '%s' and client IP '%s' successfully logged out\n", cookie.Username, clientIp)
			return
		}
	} else {
		c.AbortWithError(http.StatusUnauthorized, errors.New("user not logged in or invalid/expired cookie provided"))
		return
	}
}

// LoginFormData represents the inputs defined in the login template as a struct.
type LoginFormData struct {
	Username       string `json:"inputUsername"`
	Password       string `json:"inputPassword"`
	TOTP           string `json:"inputTotp"`
	RecaptchaToken string `json:"recaptchaToken"`
}

// RecaptchaResponse defines the structure of the Google reCAPTCHA verification response as a struct.
type RecaptchaResponse struct {
	Success            bool      `json:"success"`
	ChallengeTimestamp time.Time `json:"challenge_ts"`
	Hostname           string    `json:"hostname"`
}

// processLoginForm handles the POST /login route. If the request already contains a valid cookie, 200 is returned.
// If the user has provided valid credentials in the login form, the response will contain a new cookie (200).
// If the username, the password, the TOTP token or the reCAPTCHA token is invalid, the request is rejected.
func processLoginForm(c *gin.Context) {
	token, err := c.Cookie(GetCookieName())
	clientIp := GetClientIpFromContext(c)

	if err == nil {
		if _, err = VerifyCookie(token); err == nil {
			// user already authorized
			c.Status(200)
			return
		}
	}

	var data LoginFormData
	_ = c.Bind(&data)

	// verify reCAPTCHA token if reCAPTCHA is enabled
	if GetRecaptchaEnabled() {
		if data.RecaptchaToken == "" {
			c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"error": "bad input for reCAPTCHA token"})
			return
		}

		postBody := bytes.NewBuffer([]byte(fmt.Sprintf("secret=%s&response=%s&remoteip=%s", GetRecaptchaSecretKey(), data.RecaptchaToken, clientIp)))

		httpResponse, err := http.Post("https://www.google.com/recaptcha/api/siteverify", "application/x-www-form-urlencoded", postBody)

		if err != nil {
			c.AbortWithStatusJSON(500, gin.H{"error": "could not verify reCAPTCHA token with Google servers"})
			return
		}

		defer httpResponse.Body.Close()

		responseBody, err := io.ReadAll(httpResponse.Body)

		if err != nil {
			c.AbortWithStatusJSON(500, gin.H{"error": "could not read reCAPTCHA verification response from Google"})
			return
		}

		var response RecaptchaResponse
		err = json.Unmarshal(responseBody, &response)

		if err != nil {
			c.AbortWithStatusJSON(500, gin.H{"error": "could not deserialize reCAPTCHA verification response from Google"})
			return
		}

		if !response.Success {
			c.AbortWithStatusJSON(500, gin.H{"error": "reCAPTCHA verification unsuccessful"})
			return
		}
	}

	// prepare location data for logging purposes
	locationText := "(City: ? --- Country: ?)"
	location := GetLocationFromIP(clientIp)

	if location != nil {
		locationText = fmt.Sprintf("(City: %s --- Country: %s)", location.City, location.Country)
	}

	user := GetUserByUsername(data.Username)

	if user == nil {
		// if a user with the given username does not exist, check if LDAP authenticates
		if ldapAuthenticate(data.Username, data.Password) {
			createAndSetAuthCookie(c, data.Username)
			c.Status(200)

			authLog.Printf("LDAP user with username '%s' and client IP '%s' %s logged in successfully\n", data.Username, clientIp, locationText)
		} else {
			c.AbortWithStatus(401)
			return
		}
	} else {
		// if a user with the given username was found in the database, check password validity
		if CompareHashAndPassword(user.Password, data.Password) != nil {
			c.AbortWithStatus(401)
			authLog.Printf("invalid password for user with username '%s' %s and client IP '%s'\n", data.Username, clientIp, locationText)
			return
		} else {
			// if TOTP is enabled for the user, check the validity of the TOTP token input from the user
			if len(user.OtpSecret) != 0 {
				secret := Decrypt(user.OtpSecret, data.Password)

				tokenIsValid := totp.Validate(data.TOTP, string(secret))

				if !tokenIsValid {
					c.AbortWithStatusJSON(401, gin.H{"error": "invalid TOTP"})
					return
				}
			}

			cookie := createAndSetAuthCookie(c, user.Username)
			c.JSON(200, gin.H{"expires": cookie.Expires.UnixMilli()})
			authLog.Printf("user with username '%s' and client IP '%s' %s logged in successfully\n", data.Username, clientIp, locationText)
		}
	}
}

// createAndSetAuthCookie sets a new cookie for the given gin.Context and username and saves it to the database.
// This function is called after the user credentials have been verified.
func createAndSetAuthCookie(c *gin.Context, username string) Cookie {
	plainCookieValue := GeneratePassword(96, 25, 35)

	cookie := Cookie{
		Name:     GetCookieName(),
		Value:    GenerateHash(plainCookieValue),
		Expires:  time.Now().AddDate(0, 0, GetCookieLifetime()),
		Domain:   GetDomain(),
		Username: username,
		HttpOnly: true,
		Secure:   GetCookieSecure(),
	}

	err := SaveCookie(cookie)

	if err != nil {
		appLog.Fatalf("fatal error: could not save the cookie to the database: %s", err)
	}

	http.SetCookie(c.Writer, &http.Cookie{
		Name:     cookie.Name,
		Value:    fmt.Sprintf("$username=%s,$value=%s", username, plainCookieValue),
		Expires:  cookie.Expires,
		Domain:   cookie.Domain,
		HttpOnly: cookie.HttpOnly,
		Secure:   cookie.Secure,
	})

	SaveCookieToCache(&cookie, plainCookieValue)

	return cookie
}

// whoami handles the /whoami route. If the request contains a valid cookie,
// 200 and the username (formatted as JSON) is returned.
// If the cookie in the request header is invalid, 401 Unauthorized is returned.
func whoami(c *gin.Context) {
	token, err := c.Cookie(GetCookieName())

	if err != nil {
		c.AbortWithStatus(401)
		return
	}

	cookie, err := VerifyCookie(token)

	if err != nil {
		c.AbortWithStatus(401)
		return
	} else {
		c.JSON(200, gin.H{"username": cookie.Username})
		return
	}
}
